### PR TITLE
[posix] bind the resolver's UDP socket to the infra network interface

### PR DESCRIPTION
**Pull request by superwhd**
This PR fixes an issue in DNS recursive resolver that it didn't bind its socket to the infra network interface. This may cause the DNS message to be sent on an unexpected network interface, depending on the routing table of the platform. 

This PR also updates the test case `test_upstream_dns.py` to make the upstream DNS server run on a different node. Previously the upstream DNS server ran on the same node as the BR which is a limitation of this test case.

### CODE REVIEW

**Comment by erjiaqing** [comment]
I would suggest to make it a settings / flags.

On CPE devices, the interface used for DNS queries should be the upstream interface instead of the AIL.

Or is it possible to select / exclude a set of interfaces?

**Comment by superwhd** [comment]
> I would suggest to make it a settings / flags.
> 
> On CPE devices, the interface used for DNS queries should be the upstream interface instead of the AIL.
> 
> Or is it possible to select / exclude a set of interfaces?

Good point. I've created a build option for it.

**Review by erjiaqing — APPROVED** [review]
Thanks!

File: src/posix/platform/openthread-posix-config.h [hunk 1]
```diff
@@ -407,4 +407,13 @@
 #define OPENTHREAD_POSIX_CONFIG_RCP_CAPS_DIAG_ENABLE OPENTHREAD_CONFIG_DIAG_ENABLE
 #endif
 
+/**
+ * @def OPENTHREAD_POSIX_CONFIG_UPSTREAM_DNS_BIND_TO_INFRA_NETIF
+ *
+ * Define as 1 to let the upstream DNS bind the socket to infra network interface.
+ */
+#ifndef OPENTHREAD_POSIX_CONFIG_UPSTREAM_DNS_BIND_TO_INFRA_NETIF
+#define OPENTHREAD_POSIX_CONFIG_UPSTREAM_DNS_BIND_TO_INFRA_NETIF 0
```
Comment (jwhui) [inline src/posix/platform/openthread-posix-config.h:416]: What is the purpose of this build config? Should we always bind to the infrastructure interface?

Comment (erjiaqing) [reply src/posix/platform/openthread-posix-config.h:416]: On CPE devices, the infrastructure interface is the downlink (LAN), but we should actually send DNS requests to uplink (WAN), in this case, follow the system routing table would be a better solution.

Comment (jwhui) [reply src/posix/platform/openthread-posix-config.h:416]: I suspect most TBRs will be Wi-Fi hosts. Should we default to assume using the infrastructure interface?

Comment (superwhd) [reply src/posix/platform/openthread-posix-config.h]: updated the default to enabled

**Comment by superwhd** [comment]
I'll take a look at the failing CI.

**Review by erjiaqing — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**